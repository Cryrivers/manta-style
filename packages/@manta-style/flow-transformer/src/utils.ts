import * as ts from 'typescript';
import { types as bb } from 'babel-core';
import { MANTASTYLE_RUNTIME_NAME } from './constants';
import { QuestionToken, ReservedTypePrefix } from '@manta-style/consts';
import { identifier } from 'babel-types';

type LiteralType =
  | bb.BooleanLiteralTypeAnnotation
  | bb.NullLiteralTypeAnnotation
  | bb.NumericLiteralTypeAnnotation
  | bb.StringLiteralTypeAnnotation;

/*
type X = {
  haha: string
}

type X = {
  'haha': string
}

type X = {
  [key: string]: string
}

type X = {
  [key in Y]: string
}
*/

export function createConstVariableStatement(
  name: string,
  initializer?: bb.Expression,
) {
  return bb.variableDeclaration('const', [
    bb.variableDeclarator(bb.identifier(name), initializer),
  ]);
}

export function createTypeAliasDeclaration(node: bb.TypeAlias) {
  const { name } = node.id;
  const typeParameters: bb.TypeParameter[] =
    (node.typeParameters && node.typeParameters.params) || [];
  const jsdocTags: string[] = []; // TODO: ts.getJSDocTags(node);
  const varCreation = createConstVariableStatement(
    name,
    bb.functionExpression(
      undefined,
      undefined,
      bb.blockStatement([
        bb.returnStatement(
          createRuntimeFunctionCall('TypeAliasDeclaration', [
            bb.stringLiteral(name),
            bb.arrowFunctionExpression(
              undefined,
              bb.blockStatement([
                ...createTypeParameters(
                  typeParameters,
                  typeParameters,
                  'typeFactory',
                ),
                createConstVariableStatement(
                  'type',
                  createMantaStyleRuntimeObject(node.right, typeParameters),
                ),
                bb.returnStatement(identifier('type')),
              ]),
            ),
            generateJSDocParam(jsdocTags),
          ]),
        ),
      ]),
    ),
  );
  const registerToRuntime = createRuntimeFunctionCall('registerType', [
    bb.stringLiteral(name),
    bb.identifier(name),
  ]);
  return [varCreation, registerToRuntime];
}

export function createRuntimeFunctionCall(
  methodName: string,
  argArray: bb.Expression[],
  variableName: string = MANTASTYLE_RUNTIME_NAME,
) {
  return bb.callExpression(
    bb.memberExpression(bb.identifier(variableName), bb.identifier(methodName)),
    argArray,
  );
}

function createRuntimePropertyRef(
  propertyName: string,
  variable = MANTASTYLE_RUNTIME_NAME,
) {
  return bb.memberExpression(
    bb.identifier(variable),
    bb.identifier(propertyName),
  );
}

function createTypeParameters(
  typeParameters: bb.TypeParameter[],
  referenceTypeParameters: bb.TypeParameter[],
  caller: string,
): bb.Statement[] {
  const statements: bb.Statement[] = [];
  if (typeParameters) {
    for (const param of typeParameters) {
      statements.push(
        createConstVariableStatement(
          param.name || '',
          createRuntimeFunctionCall(
            'TypeParameter',
            [
              bb.stringLiteral(param.name || ''),
              ...(param.bound
                ? [
                    createMantaStyleRuntimeObject(
                      param.bound.typeAnnotation,
                      referenceTypeParameters,
                    ),
                  ]
                : []),
              ...(param.default
                ? [
                    createMantaStyleRuntimeObject(
                      param.default,
                      referenceTypeParameters,
                    ),
                  ]
                : []),
            ],
            caller,
          ),
        ),
      );
    }
  }
  return statements;
}

function createPropertyName(node: bb.ObjectTypeProperty) {
  const { key } = node;
  if (bb.isIdentifier(key)) {
    return ts.createStringLiteral(key.name);
  }
  throw new Error('Unsupported types when creating property name.');
}

function isGenericTypeReference(
  typeName: string,
  typeParameters: bb.TypeParameter[],
): boolean {
  return (
    (typeParameters && typeParameters.some((item) => item.name === typeName)) ||
    false
  );
}

function createTypeReferenceOrIdentifier(
  node: bb.GenericTypeAnnotation,
  typeParameters: bb.TypeParameter[],
) {
  const nodeName = node.id.name;
  if (isGenericTypeReference(nodeName, typeParameters)) {
    return ts.createIdentifier(nodeName);
  }
  return createMantaStyleRuntimeObject(node, typeParameters);
}

function generateJSDocParam(jsdocArray: string[]) {
  return bb.arrayExpression();
  // return ts.createArrayLiteral(
  //   jsdocArray.map((tag) => {
  //     return ts.createObjectLiteral([
  //       ts.createPropertyAssignment(
  //         'key',
  //         ts.createStringLiteral(tag.tagName.text),
  //       ),
  //       ts.createPropertyAssignment(
  //         'value',
  //         ts.createStringLiteral(tag.comment || ''),
  //       ),
  //     ]);
  //   }),
  // );
}

function createTypeLiteralProperties(
  members: bb.ObjectTypeProperty[],
  indexers: bb.ObjectTypeIndexer[],
  typeParameters: bb.TypeParameter[],
): bb.Statement[] {
  const statements: bb.Statement[] = [];
  for (const member of members) {
    const jsdocArray: string[] = []; //TODO: ts.getJSDocTags(member);
    statements.push(
      bb.expressionStatement(
        createRuntimeFunctionCall(
          'property',
          [
            createPropertyName(member),
            createTypeReferenceOrIdentifier(member.type, typeParameters),
            member.questionToken ? ts.createTrue() : ts.createFalse(),
            generateJSDocParam(jsdocArray),
          ],
          'typeLiteral',
        ),
      ),
    );
  }
  for (const indexer of indexers) {
    /*
    statements.push(
      bb.expressionStatement(
        createRuntimeFunctionCall(
          'computedProperty',
          [
            ts.createStringLiteral(parameter.name.getText()),
            createTypeReferenceOrIdentifier(parameter.type, typeParameters),
            createTypeReferenceOrIdentifier(member.type, typeParameters),
            ts.createNumericLiteral('0'), // ComputedPropertyOperator.INDEX_SIGNATURE
            member.questionToken ? ts.createTrue() : ts.createFalse(),
            generateJSDocParam(jsdocArray),
          ],
          'typeLiteral',
        ),
      ),
    );
    */
  }
  return statements;
}

function createBatchTypeFunctionAll(
  node: bb.UnionTypeAnnotation | bb.IntersectionTypeAnnotation,
  typeParameters: bb.TypeParameter[],
  functionName: string,
): bb.Expression {
  return createRuntimeFunctionCall(functionName, [
    bb.arrayExpression(
      node.types.map((item) =>
        createTypeReferenceOrIdentifier(item, typeParameters),
      ),
    ),
  ]);
}

function createUnionType(
  node: bb.UnionTypeAnnotation,
  typeParameters: bb.TypeParameter[],
): bb.Expression {
  return createBatchTypeFunctionAll(node, typeParameters, 'UnionType');
}

function createIntersectionType(
  node: bb.IntersectionTypeAnnotation,
  typeParameters: bb.TypeParameter[],
): bb.Expression {
  return createBatchTypeFunctionAll(node, typeParameters, 'IntersectionType');
}

function createLiteralType(node: LiteralType): bb.Expression {
  return createRuntimeFunctionCall('Literal', [bb.identifier(node.value)]);
}

function createArrayType(
  node: bb.ArrayTypeAnnotation,
  typeParameters: bb.TypeParameter[],
): bb.Expression {
  return createRuntimeFunctionCall('ArrayType', [
    createMantaStyleRuntimeObject(node.elementType, typeParameters),
  ]);
}

function createMappedType(
  node: ts.MappedTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
) {
  const referenceTypeParameters = ts.createNodeArray([
    ...typeParameters,
    node.typeParameter,
  ]);

  if (node.typeParameter.constraint && node.type) {
    return createRuntimeFunctionCall('MappedType', [
      ts.createArrowFunction(
        undefined,
        undefined,
        [
          ts.createParameter(
            undefined,
            undefined,
            undefined,
            'mappedType',
            undefined,
            undefined,
            undefined,
          ),
        ],
        undefined,
        undefined,
        ts.createBlock(
          [
            ...createTypeParameters(
              ts.createNodeArray([node.typeParameter]),
              referenceTypeParameters,
              'mappedType',
            ),
            ts.createExpressionStatement(
              createRuntimeFunctionCall(
                'setQuestionToken',
                [
                  ts.createNumericLiteral(
                    String(
                      node.questionToken
                        ? node.questionToken.kind === ts.SyntaxKind.MinusToken
                          ? QuestionToken.MinusToken
                          : QuestionToken.QuestionToken
                        : QuestionToken.None,
                    ),
                  ),
                ],
                'mappedType',
              ),
            ),
            ts.createExpressionStatement(
              createRuntimeFunctionCall(
                'setConstraint',
                [
                  createMantaStyleRuntimeObject(
                    node.typeParameter.constraint,
                    referenceTypeParameters,
                  ),
                ],
                'mappedType',
              ),
            ),
            ts.createReturn(
              createMantaStyleRuntimeObject(node.type, referenceTypeParameters),
            ),
          ],
          true,
        ),
      ),
    ]);
  } else {
    throw new Error('Invalid MappedType');
  }
}

export function createTypeLiteral(
  node: bb.ObjectTypeAnnotation,
  typeParameters: bb.TypeParameter[],
) {
  return createRuntimeFunctionCall('TypeLiteral', [
    bb.arrowFunctionExpression(
      [bb.identifier('typeLiteral')],
      bb.blockStatement(
        createTypeLiteralProperties(node.properties, typeParameters),
      ),
    ),
  ]);
}

function createTypeReference(
  node: bb.GenericTypeAnnotation,
  typeParameters: bb.TypeParameter[],
): bb.Expression {
  const typeName = node.id.name;
  // Magic type to convert query string to types
  if (typeName === 'unstable_Query' && node.typeParameters) {
    const queryKey = node.typeParameters.params[0];
    if (
      ts.isLiteralTypeNode(queryKey) &&
      ts.isStringLiteral(queryKey.literal)
    ) {
      return createRuntimeFunctionCall('TypeReference', [
        bb.stringLiteral(
          `${ReservedTypePrefix.URLQuery}${queryKey.literal.text}`,
        ),
      ]);
    } else {
      throw Error('key in unstable_Query is not a string literal.');
    }
  }
  const typeReferenceNode = isGenericTypeReference(typeName, typeParameters)
    ? ts.createIdentifier(typeName)
    : createRuntimeFunctionCall('TypeReference', [
        ts.createStringLiteral(typeName),
      ]);
  if (node.typeParameters && node.typeParameters.params) {
    return bb.callExpression(
      bb.memberExpression(typeReferenceNode, bb.identifier('argumentTypes')),
      [
        bb.arrayExpression(
          node.typeParameters.params.map((type) =>
            createTypeReferenceOrIdentifier(type, typeParameters),
          ),
        ),
      ],
    );
  } else {
    return typeReferenceNode;
  }
}

export function createMantaStyleRuntimeObject(
  node: bb.Node,
  typeParameters: bb.TypeParameter[],
): bb.Expression {
  if (bb.isObjectTypeAnnotation(node)) {
    return createTypeLiteral(node, typeParameters);
  } else if (bb.isUnionTypeAnnotation(node)) {
    return createUnionType(node, typeParameters);
  } else if (bb.isIntersectionTypeAnnotation(node)) {
    return createIntersectionType(node, typeParameters);
  } else if (ts.isLiteralTypeNode(node)) {
    return createLiteralType(node);
  } else if (bb.isArrayTypeAnnotation(node)) {
    return createArrayType(node, typeParameters);
  } else if (bb.isTupleTypeAnnotation(node)) {
    return createRuntimeFunctionCall('TupleType', [
      bb.arrayExpression(
        node.types.map((type) =>
          createMantaStyleRuntimeObject(type, typeParameters),
        ),
      ),
    ]);
  } else if (bb.isNullableTypeAnnotation(node)) {
    // TODO: This is not correct
    // should have a NullableType at runtime for Flow
    return createRuntimeFunctionCall('OptionalType', [
      createMantaStyleRuntimeObject(node.typeAnnotation, typeParameters),
    ]);
  } else if (isRestTypeNode(node) && ts.isArrayTypeNode(node.type)) {
    return createRuntimeFunctionCall('RestType', [
      createMantaStyleRuntimeObject(node.type.elementType, typeParameters),
    ]);
  } else if (ts.isIndexedAccessTypeNode(node)) {
    return createRuntimeFunctionCall('IndexedAccessType', [
      createMantaStyleRuntimeObject(node.objectType, typeParameters),
      createMantaStyleRuntimeObject(node.indexType, typeParameters),
    ]);
  } else if (ts.isConditionalTypeNode(node)) {
    return createRuntimeFunctionCall('ConditionalType', [
      createMantaStyleRuntimeObject(node.checkType, typeParameters),
      createMantaStyleRuntimeObject(node.extendsType, typeParameters),
      createMantaStyleRuntimeObject(node.trueType, typeParameters),
      createMantaStyleRuntimeObject(node.falseType, typeParameters),
    ]);
  } else if (ts.isMappedTypeNode(node)) {
    return createMappedType(node, typeParameters);
  } else if (bb.isNumberTypeAnnotation(node)) {
    return createRuntimePropertyRef('NumberKeyword');
  } else if (bb.isBooleanTypeAnnotation(node)) {
    return createRuntimePropertyRef('BooleanKeyword');
  } else if (bb.isObjectTypeAnnotation(node)) {
    return createRuntimePropertyRef('ObjectKeyword');
  } else if (bb.isStringTypeAnnotation(node)) {
    return createRuntimePropertyRef('StringKeyword');
  } else if (node.type === 'EmptyTypeAnnotation') {
    return createRuntimePropertyRef('NeverKeyword');
  } else if (bb.isNullLiteralTypeAnnotation(node)) {
    return createRuntimePropertyRef('NullKeyword');
  } else if (bb.isGenericTypeAnnotation(node) && node.id.name === 'undefined') {
    return createRuntimePropertyRef('UndefinedKeyword');
  } else if (bb.isAnyTypeAnnotation(node)) {
    return createRuntimePropertyRef('AnyKeyword');
  } else if (bb.isGenericTypeAnnotation(node)) {
    // Special Cases
    if (
      node.id.name === 'Array' &&
      node.typeParameters &&
      node.typeParameters.params.length === 1
    ) {
      return createArrayType(
        ts.createArrayTypeNode(node.typeParameters.params[0]),
        typeParameters,
      );
    }
    return createTypeReference(node, typeParameters);
  } else {
    throw new Error(`Unhandled Type. ${node.type}`);
  }
}
