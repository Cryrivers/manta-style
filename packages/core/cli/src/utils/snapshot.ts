import { HTTPMethods } from '..';
import * as fs from 'fs';
import * as path from 'path';

type AnyObject = {
  [key: string]: any;
};

type SnapshotStructure = { [key in HTTPMethods]?: AnyObject };

function snapshotKey(url: string, query: string) {
  return `${url}?${query}`;
}

export class Snapshot {
  private diskSnapshots: SnapshotStructure = {};
  private stashedSnapshots: SnapshotStructure = {};
  public static fromDisk(filePath: string) {
    const content = fs.readFileSync(path.resolve(filePath));
    const snapshotInstance = new Snapshot();
    snapshotInstance.diskSnapshots = JSON.parse(content.toString());
    snapshotInstance.stashedSnapshots = JSON.parse(content.toString());
    return snapshotInstance;
  }
  public reloadFromFile(filePath: string) {
    const content = fs.readFileSync(path.resolve(filePath));
    this.diskSnapshots = JSON.parse(content.toString());
    this.stashedSnapshots = JSON.parse(content.toString());
  }
  public updateSnapshot(
    method: HTTPMethods,
    url: string,
    query: string,
    payload: any,
  ) {
    const methodObj = this.stashedSnapshots[method];
    if (!methodObj) {
      this.stashedSnapshots = {
        ...this.stashedSnapshots,
        [method]: {
          [snapshotKey(url, query)]: payload,
        },
      };
    } else {
      this.stashedSnapshots = {
        ...this.stashedSnapshots,
        [method]: {
          ...methodObj,
          [snapshotKey(url, query)]: payload,
        },
      };
    }
  }
  public fetchSnapshot(method: HTTPMethods, url: string, query: string) {
    const methodObj = this.diskSnapshots[method];
    return methodObj && methodObj[snapshotKey(url, query)];
  }
  public clearSnapshot() {
    this.stashedSnapshots = {};
    this.diskSnapshots = {};
  }
  public writeToDisk(filePath: string) {
    const stashedJsonString = JSON.stringify(
      this.stashedSnapshots,
      undefined,
      2,
    );
    fs.writeFileSync(path.resolve(filePath), stashedJsonString);
    this.diskSnapshots = JSON.parse(stashedJsonString);
  }
}
