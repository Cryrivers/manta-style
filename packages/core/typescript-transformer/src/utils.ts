import * as ts from 'typescript';
import {
  MANTASTYLE_RUNTIME_NAME,
  MANTASTYLE_HELPER_NAME,
  MANTASTYLE_HELPER_TYPES,
} from './constants';
import { isOptionalTypeNode, isRestTypeNode } from './typescript';

const enum QuestionToken {
  None,
  QuestionToken,
  MinusToken,
}

/*
type X = {
  haha: string
}

type X = {
  'haha': string
}

type X = {
  [key: string]: string
}

type X = {
  [key in Y]: string
}
*/

export function createConstVariableStatement(
  name: string,
  initializer?: ts.Expression,
) {
  return ts.createVariableStatement(
    undefined,
    ts.createVariableDeclarationList(
      [ts.createVariableDeclaration(name, undefined, initializer)],
      ts.NodeFlags.Const,
    ),
  );
}

export function createTypeAliasDeclaration(node: ts.TypeAliasDeclaration) {
  const name = node.name.getText();
  const typeParameters = node.typeParameters || ts.createNodeArray();
  const jsdocTags = ts.getJSDocTags(node);

  const varCreation = createConstVariableStatement(
    name,
    createRuntimeFunctionCall('TypeAliasDeclaration', [
      ts.createStringLiteral(name),
      ts.createArrowFunction(
        undefined,
        undefined,
        [
          ts.createParameter(
            undefined,
            undefined,
            undefined,
            'typeFactory',
            undefined,
            undefined,
            undefined,
          ),
        ],
        undefined,
        undefined,
        ts.createBlock(
          [
            ...createTypeParameters(
              typeParameters,
              typeParameters,
              'typeFactory',
            ),
            createConstVariableStatement(
              'type',
              createMantaStyleRuntimeObject(node.type, typeParameters),
            ),
            ts.createReturn(ts.createIdentifier('type')),
          ],
          true,
        ),
      ),
      generateJSDocParam(jsdocTags),
    ]),
  );
  varCreation.modifiers = node.modifiers;
  return varCreation;
}

export function createRuntimeFunctionCall(
  methodName: string,
  argArray: ReadonlyArray<ts.Expression>,
  variableName: string = MANTASTYLE_RUNTIME_NAME,
) {
  return ts.createCall(
    ts.createPropertyAccess(
      ts.createIdentifier(variableName),
      ts.createIdentifier(methodName),
    ),
    undefined,
    argArray,
  );
}

function createRuntimePropertyRef(
  propertyName: string,
  variable = MANTASTYLE_RUNTIME_NAME,
) {
  return ts.createPropertyAccess(
    ts.createIdentifier(variable),
    ts.createIdentifier(propertyName),
  );
}

function createTypeParameters(
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
  referenceTypeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
  caller: string,
): ts.Statement[] {
  const statements: ts.Statement[] = [];
  if (typeParameters) {
    for (const param of typeParameters) {
      statements.push(
        createConstVariableStatement(
          param.name.getText(),
          createRuntimeFunctionCall(
            'TypeParameter',
            [
              ts.createStringLiteral(param.name.getText()),
              ...(param.constraint
                ? [
                    createMantaStyleRuntimeObject(
                      param.constraint,
                      referenceTypeParameters,
                    ),
                  ]
                : []),
              ...(param.default
                ? [
                    createMantaStyleRuntimeObject(
                      param.default,
                      referenceTypeParameters,
                    ),
                  ]
                : []),
            ],
            caller,
          ),
        ),
      );
    }
  }
  return statements;
}

function createPropertyName(node: ts.PropertySignature) {
  const { name } = node;
  if (ts.isStringLiteral(name)) {
    return name;
  } else if (ts.isIdentifier(name)) {
    return ts.createStringLiteral(name.getText());
  }
  throw new Error('Unsupported types when creating property name.');
}

function isGenericTypeReference(
  typeName: string,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): boolean {
  return (
    (typeParameters &&
      typeParameters.some((item) => item.name.getText() === typeName)) ||
    false
  );
}

function createTypeReferenceOrIdentifier(
  node: ts.TypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
) {
  const nodeName = node.getText();
  if (isGenericTypeReference(nodeName, typeParameters)) {
    return ts.createIdentifier(nodeName);
  }
  return createMantaStyleRuntimeObject(node, typeParameters);
}

function generateJSDocParam(jsdocArray: ReadonlyArray<ts.JSDocTag>) {
  return ts.createArrayLiteral(
    jsdocArray.map((tag) => {
      return ts.createObjectLiteral([
        ts.createPropertyAssignment(
          'key',
          ts.createStringLiteral(tag.tagName.text),
        ),
        ts.createPropertyAssignment(
          'value',
          ts.createStringLiteral(tag.comment || ''),
        ),
      ]);
    }),
  );
}

function createTypeLiteralProperties(
  members: ts.NodeArray<ts.TypeElement>,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Statement[] {
  const statements: ts.Statement[] = [];
  for (const member of members) {
    const jsdocArray = ts.getJSDocTags(member);
    if (ts.isPropertySignature(member) && member.type) {
      statements.push(
        ts.createExpressionStatement(
          createRuntimeFunctionCall(
            'property',
            [
              createPropertyName(member),
              createTypeReferenceOrIdentifier(member.type, typeParameters),
              member.questionToken ? ts.createTrue() : ts.createFalse(),
              generateJSDocParam(jsdocArray),
            ],
            'typeLiteral',
          ),
        ),
      );
    } else if (ts.isIndexSignatureDeclaration(member)) {
      if (member.parameters.length === 1 && member.type) {
        const parameter = member.parameters[0];
        if (parameter.type) {
          statements.push(
            ts.createExpressionStatement(
              createRuntimeFunctionCall(
                'computedProperty',
                [
                  ts.createStringLiteral(parameter.name.getText()),
                  createTypeReferenceOrIdentifier(
                    parameter.type,
                    typeParameters,
                  ),
                  createTypeReferenceOrIdentifier(member.type, typeParameters),
                  ts.createNumericLiteral('0'), // ComputedPropertyOperator.INDEX_SIGNATURE
                  member.questionToken ? ts.createTrue() : ts.createFalse(),
                  generateJSDocParam(jsdocArray),
                ],
                'typeLiteral',
              ),
            ),
          );
        }
      }
    }
  }
  return statements;
}

function createBatchTypeFunctionAll(
  node: ts.UnionTypeNode | ts.IntersectionTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
  functionName: string,
): ts.Expression {
  return createRuntimeFunctionCall(functionName, [
    ts.createArrayLiteral(
      node.types.map((item) =>
        createTypeReferenceOrIdentifier(item, typeParameters),
      ),
    ),
  ]);
}

function createUnionType(
  node: ts.UnionTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Expression {
  return createBatchTypeFunctionAll(node, typeParameters, 'UnionType');
}

function createIntersectionType(
  node: ts.IntersectionTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Expression {
  return createBatchTypeFunctionAll(node, typeParameters, 'IntersectionType');
}

function createLiteralType(node: ts.LiteralTypeNode): ts.Expression {
  return createRuntimeFunctionCall('Literal', [
    ts.createIdentifier(node.literal.getText()),
  ]);
}

function createArrayType(
  node: ts.ArrayTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Expression {
  return createRuntimeFunctionCall('ArrayType', [
    createMantaStyleRuntimeObject(node.elementType, typeParameters),
  ]);
}

function createMappedType(
  node: ts.MappedTypeNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
) {
  const referenceTypeParameters = ts.createNodeArray([
    ...typeParameters,
    node.typeParameter,
  ]);

  if (node.typeParameter.constraint && node.type) {
    return createRuntimeFunctionCall('MappedType', [
      ts.createArrowFunction(
        undefined,
        undefined,
        [
          ts.createParameter(
            undefined,
            undefined,
            undefined,
            'mappedType',
            undefined,
            undefined,
            undefined,
          ),
        ],
        undefined,
        undefined,
        ts.createBlock(
          [
            ...createTypeParameters(
              ts.createNodeArray([node.typeParameter]),
              referenceTypeParameters,
              'mappedType',
            ),
            ts.createExpressionStatement(
              createRuntimeFunctionCall(
                'setQuestionToken',
                [
                  ts.createNumericLiteral(
                    String(
                      node.questionToken
                        ? node.questionToken.kind === ts.SyntaxKind.MinusToken
                          ? QuestionToken.MinusToken
                          : QuestionToken.QuestionToken
                        : QuestionToken.None,
                    ),
                  ),
                ],
                'mappedType',
              ),
            ),
            ts.createExpressionStatement(
              createRuntimeFunctionCall(
                'setConstraint',
                [
                  createMantaStyleRuntimeObject(
                    node.typeParameter.constraint,
                    referenceTypeParameters,
                  ),
                ],
                'mappedType',
              ),
            ),
            ts.createReturn(
              createMantaStyleRuntimeObject(node.type, referenceTypeParameters),
            ),
          ],
          true,
        ),
      ),
    ]);
  } else {
    throw new Error('Invalid MappedType');
  }
}

export function createTypeLiteral(
  node: ts.TypeLiteralNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
) {
  return createRuntimeFunctionCall('TypeLiteral', [
    ts.createArrowFunction(
      undefined,
      undefined,
      [
        ts.createParameter(
          undefined,
          undefined,
          undefined,
          'typeLiteral',
          undefined,
          undefined,
          undefined,
        ),
      ],
      undefined,
      undefined,
      ts.createBlock(
        createTypeLiteralProperties(node.members, typeParameters),
        true,
      ),
    ),
  ]);
}

function createTypeReference(
  node: ts.TypeReferenceNode,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Expression {
  const typeName = node.typeName.getText();

  let typeReferenceNode;
  if (MANTASTYLE_HELPER_TYPES.indexOf(typeName) > -1) {
    typeReferenceNode = ts.createPropertyAccess(
      ts.createIdentifier(MANTASTYLE_HELPER_NAME),
      ts.createIdentifier(typeName),
    );
  } else {
    typeReferenceNode = ts.createIdentifier(typeName);
  }

  if (node.typeArguments) {
    return ts.createCall(
      ts.createPropertyAccess(typeReferenceNode, 'argumentTypes'),
      [],
      [
        ts.createArrayLiteral(
          node.typeArguments.map((type) =>
            createTypeReferenceOrIdentifier(type, typeParameters),
          ),
        ),
      ],
    );
  } else {
    return typeReferenceNode;
  }
}

export function createMantaStyleRuntimeObject(
  node: ts.Node,
  typeParameters: ts.NodeArray<ts.TypeParameterDeclaration>,
): ts.Expression {
  if (ts.isTypeLiteralNode(node)) {
    return createTypeLiteral(node, typeParameters);
  } else if (ts.isUnionTypeNode(node)) {
    return createUnionType(node, typeParameters);
  } else if (ts.isIntersectionTypeNode(node)) {
    return createIntersectionType(node, typeParameters);
  } else if (ts.isLiteralTypeNode(node)) {
    return createLiteralType(node);
  } else if (ts.isParenthesizedTypeNode(node)) {
    return createRuntimeFunctionCall('ParenthesizedType', [
      createMantaStyleRuntimeObject(node.type, typeParameters),
    ]);
  } else if (ts.isArrayTypeNode(node)) {
    return createArrayType(node, typeParameters);
  } else if (ts.isTupleTypeNode(node)) {
    return createRuntimeFunctionCall('TupleType', [
      ts.createArrayLiteral(
        node.elementTypes.map((type) =>
          createMantaStyleRuntimeObject(type, typeParameters),
        ),
      ),
    ]);
  } else if (isOptionalTypeNode(node)) {
    return createRuntimeFunctionCall('OptionalType', [
      createMantaStyleRuntimeObject(node.type, typeParameters),
    ]);
  } else if (isRestTypeNode(node) && ts.isArrayTypeNode(node.type)) {
    return createRuntimeFunctionCall('RestType', [
      createMantaStyleRuntimeObject(node.type.elementType, typeParameters),
    ]);
  } else if (ts.isIndexedAccessTypeNode(node)) {
    return createRuntimeFunctionCall('IndexedAccessType', [
      createMantaStyleRuntimeObject(node.objectType, typeParameters),
      createMantaStyleRuntimeObject(node.indexType, typeParameters),
    ]);
  } else if (ts.isConditionalTypeNode(node)) {
    return createRuntimeFunctionCall('ConditionalType', [
      createMantaStyleRuntimeObject(node.checkType, typeParameters),
      createMantaStyleRuntimeObject(node.extendsType, typeParameters),
      createMantaStyleRuntimeObject(node.trueType, typeParameters),
      createMantaStyleRuntimeObject(node.falseType, typeParameters),
    ]);
  } else if (ts.isMappedTypeNode(node)) {
    return createMappedType(node, typeParameters);
  } else if (node.kind === ts.SyntaxKind.NumberKeyword) {
    return createRuntimePropertyRef('NumberKeyword');
  } else if (node.kind === ts.SyntaxKind.BooleanKeyword) {
    return createRuntimePropertyRef('BooleanKeyword');
  } else if (node.kind === ts.SyntaxKind.ObjectKeyword) {
    return createRuntimePropertyRef('ObjectKeyword');
  } else if (node.kind === ts.SyntaxKind.StringKeyword) {
    return createRuntimePropertyRef('StringKeyword');
  } else if (node.kind === ts.SyntaxKind.NeverKeyword) {
    return createRuntimePropertyRef('NeverKeyword');
  } else if (node.kind === ts.SyntaxKind.NullKeyword) {
    return createRuntimePropertyRef('NullKeyword');
  } else if (node.kind === ts.SyntaxKind.UndefinedKeyword) {
    return createRuntimePropertyRef('UndefinedKeyword');
  } else if (node.kind === ts.SyntaxKind.AnyKeyword) {
    return createRuntimePropertyRef('AnyKeyword');
  } else if (
    ts.isTypeOperatorNode(node) &&
    node.operator === ts.SyntaxKind.KeyOfKeyword
  ) {
    return createRuntimeFunctionCall('KeyOfKeyword', [
      createMantaStyleRuntimeObject(node.type, typeParameters),
    ]);
  } else if (ts.isTypeReferenceNode(node)) {
    // Special Cases
    if (
      node.typeName.getText() === 'Array' &&
      node.typeArguments &&
      node.typeArguments.length === 1
    ) {
      return createArrayType(
        ts.createArrayTypeNode(node.typeArguments[0]),
        typeParameters,
      );
    }
    return createTypeReference(node, typeParameters);
  } else {
    throw new Error(`Unhandled Type. ${node.getText()}`);
  }
}
