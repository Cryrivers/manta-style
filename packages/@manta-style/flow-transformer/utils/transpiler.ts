import * as babel from 'babel-core';
import { createTransformer } from '../src';

const transformer = createTransformer(true);
export function getTranspiledString(source: string): string {
  const result = babel.transform(source, {
    plugins: [transformer],
  });
  return result.code || '';
}
