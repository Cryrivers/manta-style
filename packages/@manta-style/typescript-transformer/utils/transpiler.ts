import * as ts from 'typescript';
import { createTransformer } from '../src';

const transformer = createTransformer(true);
export function getTranspiledString(source: string): string {
  const sourceFile = ts.createSourceFile(
    'test.ts',
    source,
    ts.ScriptTarget.ES2017,
    true,
    ts.ScriptKind.TS,
  );
  const result = ts.transform(sourceFile, [transformer]);
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed,
  });
  return printer.printFile(result.transformed[0]);
}
