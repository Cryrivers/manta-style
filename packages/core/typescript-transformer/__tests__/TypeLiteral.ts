import { getTranspiledString } from '../utils/transpiler';

describe('Type Literal', () => {
  test('Basic properties', () => {
    const source = `
        type Test = {
            a: number,
            b: 1,
            c?: string,
            d: string[]
        }
      `;
    expect(getTranspiledString(source)).toMatchSnapshot();
  });
});
