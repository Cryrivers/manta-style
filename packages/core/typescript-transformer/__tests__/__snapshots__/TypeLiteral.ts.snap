// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type Literal Basic properties 1`] = `
"import MantaStyle from \\"@manta-style/runtime\\";
import * as MantaStyleHelper from \\"@manta-style/typescript-helpers\\";
const Test = MantaStyle.TypeAliasDeclaration(\\"Test\\", typeFactory => {
    const type = MantaStyle.TypeLiteral(typeLiteral => {
        typeLiteral.property(\\"a\\", MantaStyle.NumberKeyword, false, []);
        typeLiteral.property(\\"b\\", MantaStyle.Literal(1), false, []);
        typeLiteral.property(\\"c\\", MantaStyle.StringKeyword, true, []);
        typeLiteral.property(\\"d\\", MantaStyle.ArrayType(MantaStyle.StringKeyword), false, []);
    });
    return type;
}, []);
"
`;
